lib/translator/inquiry.cc

```diff
@@ -0,0 +1,292 @@
+  result.page_code = 0x80;
+
+  // TODO: write this after scsi response in buffer.
+  char *product_serial_number[100];
+
+  // check if nonzero
+  bool nguid_nz =
+      identify_namespace_data.nguid[0] || identify_namespace_data.nguid[1];
+  bool eui64_nz = identify_namespace_data.eui64;
+  if (eui64_nz) {
+    if (nguid_nz) {
+      // 6.1.3.1.1
+      result.page_length = 40;
+
+      // convert 128-bit number into hex string, 64-bits at a time
+      char hex_string[33];
+      sprintf(hex_string, "%08lx", identify_namespace_data.nguid[0]);
+      sprintf(&hex_string[16], "%08lx", identify_namespace_data.nguid[1]);
+
+      // insert _ and . in the correct positions
+      char formatted_hex_string[41];
+      for (int i = 4; i < result.page_length; i += 5) {
+        formatted_hex_string[i] = '_';
+      }
+      formatted_hex_string[result.page_length - 1] = '.';
+
+      // insert numbers
+      int pos = 0;
+      for (int i = 0; i < result.page_length - 1; i++) {
+        if (formatted_hex_string[i] != '_') {
+          formatted_hex_string[i] = hex_string[pos++];
+        }
+      }
+
+      memcpy(product_serial_number, formatted_hex_string, result.page_length);
+    } else {
+      // 6.1.3.1.2
+      result.page_length = 20;
+
+      // convert 64-bit number into hex string
+      char hex_string[17];
+      sprintf(hex_string, "%08lx", identify_namespace_data.eui64);
+
+      // insert _ and . in the correct positions
+      char formatted_hex_string[21];
+      for (int i = 4; i < result.page_length; i += 5) {
+        formatted_hex_string[i] = '_';
+      }
+      formatted_hex_string[result.page_length - 1] = '.';
+
+      // insert numbers
+      int pos = 0;
+      for (int i = 0; i < result.page_length - 1; i++) {
+        if (formatted_hex_string[i] != '_') {
+          formatted_hex_string[i] = hex_string[pos++];
+        }
+      }
+
+      memcpy(product_serial_number, formatted_hex_string, result.page_length);
+    }
+  } else {
+    if (nguid_nz) {
+      // 6.1.3.1.1
+      result.page_length = 40;
+
+      // convert 128-bit number into hex string, 64-bits at a time
+      char hex_string[33];
+      sprintf(hex_string, "%08lx", identify_namespace_data.nguid[0]);
+      sprintf(&hex_string[16], "%08lx", identify_namespace_data.nguid[1]);
+
+      // insert _ and . in the correct positions
+      char formatted_hex_string[41];
+      for (int i = 4; i < result.page_length; i += 5) {
+        formatted_hex_string[i] = '_';
+      }
+      formatted_hex_string[result.page_length - 1] = '.';
+
+      // insert numbers
+      int pos = 0;
+      for (int i = 0; i < result.page_length - 1; i++) {
+        if (formatted_hex_string[i] != '_') {
+          formatted_hex_string[i] = hex_string[pos++];
+        }
+      }
+
+      memcpy(product_serial_number, formatted_hex_string, result.page_length);
+    } else {
+      // 6.1.3.1.3
+      // valid for NVMe 1.0 devices only
```

Comment:
Add a DebugLog or something else to notify or a failure/unexpeted input. Ideally we would want to be able to return a status code in these scenarios, so the caller can react in case of failure.